pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,664 @@
+		},
+		tryToCreate: func(ctx context.Context, userName string, logicalClusterName logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryAfterSeconds int, createError error) {
+			return tryToCreate(h, ctx, userName, logicalClusterName, workspaceType)
+		},
+	}
+
+	return h
+}
+
+func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if !h.localInformers.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	lcluster, err := request.ValidClusterFrom(ctx)
+	if err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	if sets.NewString(user.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+		// If we are the system privileged group, it might be a call from the virtual workspace
+		// in which case we also search the user in the soft impersonation header.
+		if impersonated, err := softimpersonation.UserInfoFromRequestHeader(req); err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		} else if impersonated != nil {
+			user = impersonated
+		}
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
+		return
+	}
+
+	if lcluster.Name == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+
+		// if we are in the special case of a `kubectl get workspace ~` request (against the 'root' workspace),
+		// we return the Home workspace definition of the current user, possibly even before its
+		// underlying ClusterWorkspace resource exists.
+
+		homeLogicalClusterName := h.getHomeLogicalClusterName(user.GetName())
+		homeClusterWorkspace, err := h.localInformers.getClusterWorkspace(homeLogicalClusterName)
+		if err != nil && !kerrors.IsNotFound(err) {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he requests to see its definition.
+			// We don't need to check any permission before returning the home workspace definition since,
+			// once it has been created, a home workspace is owned by the user.
+
+			homeWorkspace := &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+			responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+			return
+		}
+
+		// Test if the user has the right to see his Home workspace even though it doesn't exists
```

Comment:
If the user always has a right to see their home workspace, once it exists, what does it mean to check their permissions here? In what case would this check fail?

Reply:
We're precisely in the case when the home workspace doesn't exist yet.
We manage a dedicated permission for:

*being able to see where the home workspace will be (== what it's logical cluster will be) even _**before it exists**_*

When the Home Workspace doesn't not exist, we want to gate the ability:
- to know where the Home workspace will be
- to create the HomeWorkspace automatically on first access.

The 2 are granted by default for authenticated users.

See this already-merged PR for more details: https://github.com/kcp-dev/kcp/pull/1455/files#diff-e4317fcacf05b074783ba5e267a2d0fafd48d0069d9fb8ed0e453b39f3bbb2a8

Reply:
I think the gate here is fine and expected. I wonder though whether we want additional authz for the other case as well, not in the sense of gating, but just to ensure it is really the workspace of this user and nobody else due to a conflict.